### PR TITLE
Allow table columns to be hidden from print output

### PR DESCRIPTION
**Pull request by taldcroft**
### Description

This pull request is another try at #2466 to allow table columns to be hidden during normal print functions.

This now stores the hidden column information on the table, not on columns, thus addressing one original concern from @mhvk in #2466.

It adds a handy context manager to allow unhiding all columns in the context block.

This implementation stores  the hidden column set as a Table MetaAttribute, i.e. in `t.meta['__attributes__']`. This allows all the usual machinery like slicing and serialization to work since `meta` is guaranteed to propagate.

The API is intentionally similar to `ConfigItem`, with methods `__call__` (to get the value), and `set` for setting, `set_temp` (context manager).

## Demonstration code
```
import astropy
from astropy.table.table_helpers import simple_table
from io import StringIO
from astropy.io import ascii
print(astropy.__version__)

t = simple_table(cols=6)
print(t.hidden_columns)
# <HiddenColumns name=hidden_columns value=None>

print(t)

t.hidden_columns = ('c', 'd', 'f')

# Same thing with the `set` method ala ConfigItem
t.hidden_columns.set(('c', 'd', 'f', 'does-not-exist'))  # Can set non-existent col

# __call__ provides the actual value as a tuple (ala ConfigItem)
cols = t.hidden_columns()

print(t.hidden_columns)
# <HiddenColumns name=hidden_columns value=('c', 'd', 'f', 'does-not-exist')>

print(t)

# Serialization works
out = StringIO()
ascii.write(t, out, format='ecsv')
print(out.getvalue())
t2 = ascii.read(out.getvalue(), format='ecsv')
print(t2)

# Context manager ala ConfigItem
with t2.hidden_columns.set_temp(None):
    print(t2)

with t2.hidden_columns.set_temp('b'):
    print(t2)

t2.hidden_columns.remove('c')
```

### CODE REVIEW

**Comment by pllim** [comment]
CI skip won't work until #11168 is merged. FYI

**Review by mhvk — COMMENTED** [review]
Wow, that is not much extra code for a nice feature! It does seem good on the table.

On the implementation/API, I wonder if it might not be possible to make `hide_columns` the descriptor (so that the context manager would not temporarily unhide some, but just hide some).

I even wondered if it could be all combined in a `hidden_columns` descriptor, but that seems tricky.  One could consider that having a setter, though.

File: astropy/utils/metadata.py
Comment (taldcroft) [reply astropy/utils/metadata.py:459]: > and if it is set to the default, any existing entry is deleted.

I'm not sure where you are imagining deleting an existing entry if it is set to the default, but in any case I can't see how to do that without doing an equality comparison of the existing entry to the default. But generally speaking equality comparisons of arbitrary user-supplied objects are dangerous.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:360]: I was coming at this from the use case that both @Cadair and I have had, namely a Table subclass that has a known list of columns that are typically not interesting. This does beg the question of whether it should be allowed to hide columns that don't (yet) exist. That would allow putting a `self.hide_columns(not_interesting_columns)` into the subclass `__init__`. This seems like something that would be useful.

I see your point that the converse (HiddenColumns) could be useful as well. I need to think about this more to see if it is possible. The exact structure for `UnhideColumnsMethod` won't work but there might be a variation...

Worst case is being less fancy with:
```
t.hide_columns(cols)
t.unhide_columns(cols)
with t.hide_columns_context(cols):
     ...
with t.unhide_columns_context(cols):
    ...
```

**Review by taldcroft — COMMENTED** [review]
Thanks for the useful feedback @mhvk!  I'll continue thinking about this.

**Comment by taldcroft** [comment]
I have a different idea that might work.. going to give it a try.

**Comment by taldcroft** [comment]
@mhvk - so now there is only the `hidden_columns` attribute, which is a `HiddenColumn` descriptor that inherits from `MetaAttribute` but adopts the API of `ConfigItem`. I've updated the description with new demonstration code. Let me know if this seems OK or if you have comments. If I have a green light on this I'll finish this up with tests and docs and code comments.

**Comment by taldcroft** [comment]
@mhvk - oops, there is a fundamental problem now with the implementation. It is storing `_instance_ref` on the descriptor `self` instead of on the `instance`.  I need to use some of the trickery in `DataInfo` to make this work.

**Comment by taldcroft** [comment]
One thing is that my own use case for this makes more sense specifying the interesting columns to show, not the boring ones to hide. I've been thinking about this and wondering about generalizing the current `hidden_columns` attribute to go either way: list of hidden columns or a list of shown columns.

File: astropy/table/table.py
Comment (mhvk) [inline astropy/table/table.py:416]: With `add` and `remove`, one gets part of the behaviour of a set, but not all.  An option might be to actually store a `set` inside and then do something like:
```
def __getattr__(self, attr):
    value = self()
    if hasattr(value, attr):
        return getattr(value, attr)
```
Then even things like `table.hidden_columns |= {'a'}` would work...
(Or possibly one can be a mixin of `MetaAttribute` and `set`??)

Note: obviously, *not* important - since that would just make `.add` and `.remove` work the same way as they are now.

File: astropy/table/table.py
Comment (mhvk) [inline astropy/table/table.py:304]: I think you only need to override `__delitem__` - `pop` and `popitem` end up there as well.

File: astropy/table/table.py
Comment (mhvk) [inline astropy/table/table.py:302]: Don't you need to check that the `name` was hidden before? I think `.remove` would give an error.

File: astropy/table/table.py
Comment (mhvk) [inline astropy/table/table.py:427]: Following `set()`, `KeyError` may be more appropriate.

**Review by mhvk — COMMENTED** [review]
I like this! Quite a few comments inline, but all implementation details.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py]: I actually started with that then discovered this interesting behavior with `KeyError` that put me off. It displays the `repr` of the argument, not the `str`. I think somehow the intent is that you should raise `KeyError(key)`, which doesn't quite fit here. But I can go either way.
```
In [10]: raise ValueError('my message')                                                                                             
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-a8f0b130d67a> in <module>
----> 1 raise ValueError('my message')

ValueError: my message

In [11]: raise KeyError('my message')                                                                                               
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-11-0aef9f386bb7> in <module>
----> 1 raise KeyError('my message')

KeyError: 'my message'
```

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py]: Yup, that is fixed in the latest version.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:302]: Cool, will do.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:450]: The external API and implementation are more like `tuple` now instead of `set`. The `add` and `remove` method names are following the general schema of `Table` methods, not `set`.

One benefit of tuple, which may be only a benefit to my own version of tidiness, is that the YAML repr of a set is ugly (`!!set {a:null, b:null, c:null}`).

But I agree there is something compelling to the idea of `set` underneath, so I'll think about it some more.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py]: Hah, I saw the original code this morning (before seeing your comments) and did exactly that!

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:416]: Much of this is lesson-learned from `DataInfo` without re-thinking things, but in general I think avoiding direct circular refs is a good thing.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:432]: > Would it make sense to add or () at the end so that one always gets a tuple?

This would break the nice implementation of `set_temp()` because it would morph an original value of `None` to a final value of `()`.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:423]: Another instance where I made the same change (re-reviewing my code this morning) before seeing your comment. :smile:

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:484]: Overtaken by adding a separate `columns_show` attribute. BTW, the advantage there is that it allows setting the "interesting" columns in advance whether or not they exist yet.
```
class MyTable(Table):
    def __init__(*args, **kwargs):
        super().__init__(*args, **kwargs)
        self.columns_show = ('a', 'b', 'c')
```

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:421]: True, but I prefer to have argument munging up top as a distinct step for code clarity. Performance of an extra code line is not an issue here.

**Review by taldcroft — COMMENTED** [review]
Thanks for the great review, I'm going to press ahead to finish this up. About the internals and possibly exposing a `set` object, it seems like what's there is sufficient. Astropy Table has all these `add_*` and `remove_*` methods, and I don't see a real need for users to have the full power of sets in the `columns_show/hide` attributes. I think the full user API is pretty much just this:
```
t.columns_hide = ['a', 'b']  # Set one or more columns
t.columns_hide = None  # Clear them
t.columns_hide.add/remove(['c', 'd'])  # Add/remove columns
with t.columns_hide.set_temp(['c', 'd']):  # Context manager to hide 'c', 'd'
with t.columns_hide.set_temp(None):  # Context manager to unhide all columns
```
One question is whether `columns_show` should be more like `columns_show_only`, which is really more descriptive of what's happening.

**Comment by mhvk** [comment]
Hmm, I worry about having two separate items keeping track of something related. For instance, currently if you set an overall `table.columns_hide = 'a'`, and then try `with table.columns_show.set_temp('a')`, I think it will not actually show 'a'.

If the main use case is selectively showing columns, perhaps that should be the only descriptor, and it should also have a `hide` keyword argument? It could be called `column_visibility` to ensure the name is generic enough that it suggests one can hide too.

Also, maybe as an extension, should the entries be treated as regex? Many-column tables often have series with similar names...

**Comment by taldcroft** [comment]
@mhvk - I was also not happy with the two-attribute solution, and spent a fair bit of time trying to think of a one-attribute solution which is clean . Unfortunately I failed, but I'm open to suggestions. The problem is that you need to carry two distinct pieces of information: the "action" (show or hide) and the list of names related to the action. The API then needs to reflect this `action` everywhere, and I think that just gets more messy and less explicit (since there is this implicit `action` property that gives the meaning to the list of column names).

I don't think there is an obvious dominant use case. It's basically like any number of tools e.g. `tar` that have --include and --exclude options. It is always a bit awkward and you just have to document the rule, which in this case is that `show` gets evaluated first and `hide` second. Thus there is no ambiguity and no surprises.

In practice I don't think users will often combine the two, so you will just see code with one of `show` or `hide` and the meaning will be obvious.

**Comment by taldcroft** [comment]
@mhvk - about the regex, I would suggest just `fnmatch` (globbing) as being enough. In particular `.` is probably pretty common in column names (causing unexpected results for some users) while `*` `?` `[]` are not so common. It's easy enough to code, the question is whether it adds too much complexity. I can go either way.

**Comment by taldcroft** [comment]
BTW, I was just looking in `io.ascii` and was reminded of `include_names` and `exclude_names` there. Maybe it would be more clear if the attributes were `pprint_include_names` and `pprint_exclude_names`?  Then `t.pp<TAB>` would show them all.

**Comment by mhvk** [comment]
You're right that typically people would not do both and I see the case for preventive setting of names that do not yet exist. I like the analogy with `include_names` and `exclude_names` - in the end this is just output to stdout after all! One could then think of adding those to `pprint` itself too...

**Comment by mhvk** [comment]
Indeed, just the ability to add a `*` would be enough, I'd think...

**Comment by taldcroft** [comment]
@mhvk - I think this is ready for final review. Tests are passing, docs added.

**Comment by taldcroft** [comment]
Probably when this is approved I'll squash it down since the development path was a bit circuitous.

File: astropy/table/table.py
Comment (mhvk) [inline astropy/table/table.py:534]: Just a thought: `astropy.utils.state.ScienceState.set()` returns a context so that it can be used as a context manager if need be. I can see the benefit of just sticking with the `ConfigItem` analogy, though.

**Review by mhvk — APPROVED** [review]
Looks good! One thing just to be sure we considered it, otherwise 2 nitpicks which you might as well do when squashing.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py:534]: Good idea with the `ConfigItem` analogy. I had to convince myself that the dual setting and context manager functionality is bombproof, but it seems to be. I like the API somewhat better.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py]: Done.

File: astropy/table/table.py
Comment (taldcroft) [reply astropy/table/table.py]: Done.

**Pull request merged**
